Adaround Implementation

edmundw314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41881 Adaround Implementation**

Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

edmundw314 committed:
Adaround Implementation

Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_adaround.py:
```diff
@@ -0,0 +1,57 @@
+import torch.nn as nn
+from torch.testing._internal.common_quantization import QuantizationTestCase
+from torch.quantization import QuantStub, DeQuantStub
+from torch.quantization._adaround import _supported_modules
```
lint

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+    else:
+        return split_name[0], split_name[1]
+
+class adaround(FakeQuantize):
```
nit: AdaRound for class name

jerryzh168 replied:
maybe `AdaRoundFakeQuantize` to make it more clear that this is a fake quantize module

jerryzh168 replied:
also this should probably be in a different file

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+            # small initial values are intended so no bias is introduced on initialization or calibration
+            self.continous_V = torch.nn.Parameter(torch.ones(X.size()) / 10000)
+
+        if self.observer_enabled[0] == 1:
```
can you add a comment on why we are not reusing the forward of parent module? I already forgot what reason is actually

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+                                                          self.quant_max)
+        return X
+
+    def randomize(self):
```
feel free to remove if this is not used

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+        return self.dequant(self.quantized_output)
+
+
+araround_fake_quant = adaround.with_args(observer=HistogramObserver, quant_min=-128, quant_max=127,
```
default_adaround_fake_quant?

jerryzh168 replied:
btw, using HistogramObserver might be the reason that it is slow, you can try MinMax to make it faster when you run tests.

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+_supported_modules = {nn.Conv2d, nn.Linear}
+
+# Hyper parameters for loss function
+beta_high = 8
```
for constant we typically use all capitable case, e.g. BETA_HIGH = 8

jerryzh168 replied:
also some of these parameters should probably be configurable in the adaround module I think

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+regularization_scaling = .1
+
+# Adaround training parameters
+number_of_epochs = 10
+number_of_calibration_batches = 10
+learning_rate = .1
```
these should probably be argument for learn_adaround function

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+            self.zero_point = _zero_point
+
+        if self.fake_quant_enabled[0] == 1:
+            X = adaround_rounding(self, X)
```
why is this not a method of the adaround fake quantize module?

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+    '''
+    result = torch.Tensor([0])
+    for name, submodule in model.named_modules():
+        if isinstance(submodule, OuputWrapper):
```
do we need this wrapper? can you just check the type of the module directly? if it is AdaRoundFakeQuantize module then we will accumulate the loss

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+    weights_w_adaround_rounding = scale * torch.clamp(weights_clipped, model.quant_min, model.quant_max)
+    return weights_w_adaround_rounding
+
+def loss_function_leaf(model, count):
```
this can probably also be a part of the AdaRoundFakeQuantize module

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+    for name, submodule in model.named_modules():
+        if isinstance(submodule, OuputWrapper):
+            result = result + loss_function_leaf(submodule, count)
+            result = result * .90
```
what is this?

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+            result = result * .90
+    return result
+
+def add_wrapper_class(model, white_list=DEFAULT_QAT_MODULE_MAPPING.keys()):
```
nit: white_list --> supported_list

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+            submodule.qconfig = adaround_qconfig
+
+
+def prepare_adaround(float_model, dataset, white_list=_supported_modules):
```
same here, change white_list

jerryzh168 replied:
users are supposed to attach qconfig themselves in eager mode, I don't think we should call prepare_qat for user. AdaRoundFakeQuantize module will just be a module that user can use in replacement of FakeQuantize module. we should implement loss functions etc. as a method of AdaRoundFakeQuantize module itself.

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+            parent_name, child_name = parent_child_names(name)
+            parent_module = get_module(model, parent_name)
+            parent_module._modules[child_name] = OuputWrapper(submodule)
+            submodule.qconfig = adaround_qconfig
```
these things should be configurable by user I think

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,260 @@
+        init.kaiming_uniform_(self.continous_V, a=math.sqrt(5))
+
+
+class OuputWrapper(nn.Module):
```
can we do this in the forward of adaround fake quantize module?

jerryzh168 reviewed (changes requested):
I think in general we should move most of the logic to AdaRoundFakeQuantize module

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_adaround.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class TestAdaround(QuantizationTestCase):
+    def single_layer_adaround(self, model, adaround_func, img_data):
```
this is not implemented yet

I think we can compare with default fake quant and check if the sqnr is always better

raghuramank100 replied:
We should compare with default PTQ, not fake-quant.

jerryzh168 replied:
how is default fake quant without learning weight different from PTQ?

jerryzh168 replied:
@edmundw314 pleases make sure when you compare with default fake quant do not make the weight learnable. it should just be observing and fake quant, but without backprop or the optimizer should not have any parameters.

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/adaround_fake_quantize.py:
```diff
@@ -0,0 +1,102 @@
+        settings = AdaRoundFakeQuantize.adaround_default_attribute_settings
+
+        for attribute in settings:
+            if attribute in settings:
```
you mean `if attribute in keywords`?

jerryzh168 commented on torch/quantization/adaround_fake_quantize.py:
```diff
@@ -0,0 +1,102 @@
+        for attribute in settings:
+            if attribute in settings:
+                setattr(self, attribute, keywords[attribute])
+                del keywords[attribute]
```
why do we need to delete here?

edmundw314 replied:
keywords gets passed to its parent class FakeQuantize, i thought weird things would happen if im passing extra parameters to FakeQuantize

jerryzh168 replied:
you should call super first with explicit arguments and then initialize the remaining attributes

jerryzh168 commented on torch/quantization/adaround_fake_quantize.py:
```diff
@@ -0,0 +1,102 @@
+                                           'norm_scaling': 10,
+                                           'regularization_scaling': .1}
+
+    def __init__(self, *args, **keywords):
```
I feel it might be better to just spell out the arguments in the list, also please add a comment for all the arguments

jerryzh168 commented on torch/quantization/adaround_fake_quantize.py:
```diff
@@ -0,0 +1,102 @@
+        super(AdaRoundFakeQuantize, self).__init__(*args, **keywords)
+
+    def forward(self, X):
+        if self.continous_V is None:
+            # need some value already assigned to continuous_V for kaiming to work
+            self.continous_V = torch.nn.Parameter(torch.zeros(X.size()))
+            init.kaiming_uniform_(self.continous_V, a=math.sqrt(5))
```
this should be done in __init__?

edmundw314 replied:
can't be done in init because you need to know X.size() (which is unknown when init is called), to get around that i set continous_V to None and the first time the forward is run we get X.size() and its updated accordingly

jerryzh168 replied:
oh, I see, please add a comment

jerryzh168 commented on torch/quantization/adaround_fake_quantize.py:
```diff
@@ -0,0 +1,102 @@
+            self.continous_V = torch.nn.Parameter(torch.zeros(X.size()))
+            init.kaiming_uniform_(self.continous_V, a=math.sqrt(5))
+
+        if self.tuning:
```
`tuning` --> `enable_adaround`

jerryzh168 commented on torch/quantization/adaround_fake_quantize.py:
```diff
@@ -0,0 +1,102 @@
+
+        return Frobenius_norm * self.norm_scaling + self.regularization_scaling * regulization
+
+parameters = {'observer': MovingAverageMinMaxObserver, 'quant_min': -128, 'quant_max': 127, 'dtype': torch.qint8,
```
these can be removed when you have default argument

jerryzh168 commented on torch/quantization/_adaround.py:
```diff
@@ -0,0 +1,81 @@
+        weight_fake_quant attribute and runs an adam optimizer on the continous_V
+        attribute on the adaround module
+        '''
+        leaf_module.weight_fake_quant.tuning = True
```
is this needed?

edmundw314 replied:
weird things happened when i didn't separate the training for continous_V from the normal qat calibration, if we get rid of this the qat logic has trouble finding good values for scale and offset i think (the accuracy in demos plummets to ~15-17% when we don't add this)

jerryzh168 replied:
oh, are you using the same model to do the comparison?

jerryzh168 reviewed (commented):
see inline comments

edmundw314 committed:
Update on "Adaround Implementation"


Summary: Adding FakeQuant to implement a new clipping procedure to improve accuracy of
quantization.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22676684](https://our.internmc.facebook.com/intern/diff/D22676684)

[ghstack-poisoned]
